pkg/defaults/defaults.go

```diff
@@ -787,8 +763,66 @@ func stepConfigsForBuild(
 	}
 
 	buildSteps = append(buildSteps, config.RawSteps...)
+	return api.GraphConfiguration{Steps: buildSteps}
+}
 
-	return api.GraphConfiguration{Steps: buildSteps}, nil
+func runtimeStepConfigsForBuild(
+	ctx context.Context,
+	client ctrlruntimeclient.Client,
+	config *api.ReleaseBuildConfiguration,
+	jobSpec *api.JobSpec,
+	readFile readFile,
+	resolveRoot resolveRoot,
+	imageConfigs *[]*api.InputImageTagStepConfiguration,
+	second time.Duration,
+) ([]api.StepConfiguration, error) {
+	var buildSteps []api.StepConfiguration
+	if root := config.InputConfiguration.BuildRootImage; root != nil {
+		var target *api.InputImageTagStepConfiguration
+		if root.FromRepository || root.UseBuildCache {
+			for i, s := range *imageConfigs {
+				if to := s.InputImage.To; to == api.PipelineImageStreamTagReferenceRoot {
+					target = (*imageConfigs)[i]
+					break
+				}
+			}
+		}
+		if target != nil {
+			istTagRef := &target.InputImage.BaseImage
+			if root.FromRepository {
+				var err error
+				istTagRef, err = buildRootImageStreamFromRepository(readFile)
+				if err != nil {
+					return nil, fmt.Errorf("failed to read buildRootImageStream from repository: %w", err)
+				}
+				ensureImageStreamTag(ctx, client, istTagRef, second)
```

Comment:
Is this a valid refactor? Looks like we're ensuring the input image before doing the build cache override?

Reply:
Technically it is, because it maintains what we had before.

But the sequence does seem suspicious, I'll take a look.

Reply:
https://github.com/openshift/ci-tools/pull/2242